### PR TITLE
feat: workflow services — expose workflows as REST microservice endpoints

### PR DESCRIPTION
**Pull request by edurdias**
## Overview

Adds **workflow services** — a way to expose Flux workflows as clean REST microservice endpoints. A service is a named collection of workflow selectors (namespaces and/or individual workflows, minus exclusions) that dynamically resolves to a set of HTTP endpoints. Each workflow in the service becomes a POST endpoint where the request body is the workflow input and the response body is the workflow output.

Services resolve selectors at request time — registering a new workflow in a namespace that a service selects automatically makes it available as an endpoint, with no restart or reconfiguration.

When auth is enabled, service endpoints inherit the same permission checks as direct workflow execution (`workflow:{namespace}:{name}:run`). When auth is disabled (default), service endpoints are fully open.

### Service Definition (selector-based, dynamic)
- A service has a **name**, **namespace selectors** (include all workflows from these namespaces), **individual workflow refs** (include specific qualified refs like `billing/invoice`), and **exclusions** (remove specific workflows from the resolved set).
- Resolution happens at request time against the live `WorkflowCatalog`. Adding or removing workflows from the catalog instantly updates the service's available endpoints.
- **Name collision detection**: if two included workflows from different namespaces share the same name, the service returns 409 Conflict. Exclusions can resolve collisions by removing one side.
- Services are stored in a `services` table with JSON columns for selectors.

### Built-in Mode (zero config)
- Service endpoints are served on the main Flux server at `/services/{service_name}/{workflow_name}`.
- The request body becomes the workflow input. The response body is the workflow output (raw by default, or wrapped with execution metadata via `?detailed=true`).
- Supports `sync` (default), `async`, and `stream` execution modes via `?mode=` query param or `/{mode}` path segment.
- HTTP status mapping: COMPLETED → 200, FAILED → 500 with error, PAUSED → 202 with resume URL, async → 202 with status URL.

### Standalone Mode (independent deployment)
- `flux service start billing --port 9000 --server-url http://flux:8000` spawns a dedicated FastAPI process serving a single service at root-level URLs (`POST /invoice`, `POST /refund`).
- The proxy forwards requests to the main Flux server via HTTP, passing through `Authorization` headers.
- **Lazy endpoint discovery with cache**: the proxy fetches the service's endpoint list from the server, caches it with a configurable TTL (default 60s), and auto-discovers new workflows on cache miss. Lock-protected refresh prevents thundering herd under load.
- Health endpoint at `GET /health` reports service name, endpoint count, and cache age.

### Response Behavior

Default (sync) returns raw workflow output:
```json
{"invoice_id": "INV-123", "total": 99.50}
```

With `?detailed=true`:
```json
{
  "data": {"invoice_id": "INV-123", "total": 99.50},
  "execution_id": "abc123",
  "state": "COMPLETED",
  "workflow_namespace": "billing",
  "workflow_name": "process_payment"
}
```

Async returns 202 with status URL:
```json
{
  "execution_id": "abc123",
  "state": "CREATED",
  "status_url": "/services/billing/process_payment/status/abc123"
}
```

Paused returns 202 with resume URL:
```json
{
  "execution_id": "abc123",
  "state": "PAUSED",
  "resume_url": "/services/billing/process_payment/resume/abc123"
}
```

## CLI

New top-level `flux service` group:
- `create <name> [--namespace NS]... [--workflow REF]... [--exclude REF]...` — create a service with selectors
- `show <name>` — display service definition and resolved endpoints
- `list` — list all services
- `add <name> [--namespace NS]... [--workflow REF]...` — add selectors to existing service
- `remove <name> [--namespace NS]... [--workflow REF]...` — remove selectors
- `exclude <name> <refs>...` — exclude specific workflows from the service
- `include <name> <refs>...` — un-exclude workflows (restore them)
- `delete <name> [--yes]` — delete a service
- `start <name> [--port PORT] [--server-url URL] [--cache-ttl SECONDS]` — start standalone proxy

All commands support `--format json` and `--server-url`.

## Server API Endpoints

```
POST   /services                                           — Create service
GET    /services                                           — List services
GET    /services/{name}                                    — Show service + resolved endpoints
PUT    /services/{name}                                    — Update selectors
DELETE /services/{name}                                    — Delete service
POST   /services/{name}/{workflow}[/{mode}]                — Run workflow via service
POST   /services/{name}/{workflow}/resume/{exec_id}[/{mode}] — Resume paused workflow
GET    /services/{name}/{workflow}/status/{exec_id}        — Check execution status
```

## New Files

| File | Purpose |
|---|---|
| `flux/service_store.py` | Service CRUD — `ServiceModel`, `ServiceInfo`, `ServiceStore`, `ServiceNotFoundError` |
| `flux/service_resolver.py` | Dynamic selector resolution — `ServiceResolver`, `CollisionError`, `WorkflowNotInServiceError` |
| `flux/service_proxy.py` | Standalone proxy — `StandaloneServiceProxy`, `create_standalone_app()` |
| `flux/cli_service.py` | CLI commands — 9 subcommands under `flux service` |

## Validation

- **23 unit tests** for `ServiceStore` CRUD (create, get, list, update edge cases, delete)
- **10 unit tests** for `ServiceResolver` (namespace/workflow selectors, exclusions, collision detection/resolution, empty services, mixed selectors)
- **10 E2E tests** against real server+worker:
  - Service CRUD lifecycle (create → show → list → delete)
  - Add/remove/exclude/include selectors
  - Sync endpoint execution with correct output
  - Detailed response mode with execution metadata
  - Async mode returns 202 with status URL
  - 404 for nonexistent workflow or service
  - Dynamic discovery (register workflow after service creation → appears)
  - Exclusion enforcement (excluded workflow → 404)
  - Status endpoint via service
  - Workflow deletion reflected in service (→ 404)
- **All existing tests unaffected** (1632 unit tests pass, CI green)

### CODE REVIEW

File: flux/service_proxy.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                return
+            response = await self._client.get(f"/services/{self.service_name}")
+            response.raise_for_status()
+            data = response.json()
+            endpoints: dict[str, WorkflowRef] = {}
+            for ep in data.get("endpoints", []):
+                parts = ep.split("/", 1)
+                if len(parts) == 2:
+                    ref = WorkflowRef(namespace=parts[0], name=parts[1])
+                    endpoints[parts[1]] = ref
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py]: Fixed in 29859c7. Proxy cache now parses dict format (`{"name": ..., "namespace": ...}`) from server response, with string fallback.

File: flux/service_resolver.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                    )
+                else:
+                    workflows[wf.name] = wf
+
+        for ref in service.exclusions:
+            ns, name = resolve_workflow_ref(ref)
+            if name in workflows and workflows[name].namespace == ns:
+                del workflows[name]
+            collisions.pop(name, None)
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_resolver.py]: Fixed in 29859c7. Resolver now uses candidates-based approach — tracks all candidates per name, applies exclusions by full ref, then detects collisions if >1 candidate remains.

File: flux/server.py [hunk 3]
```diff
@@ -2958,6 +2958,557 @@ async def schedule_history(
...
+
+            body = await request.json()
+            try:
+                store = ServiceStore()
+                svc = store.create(
+                    name=body["name"],
+                    namespaces=body.get("namespaces", []),
+                    workflows=body.get("workflows", []),
+                    exclusions=body.get("exclusions", []),
+                )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py]: Fixed in 29859c7. Added validation: returns 400 if name is missing or not a string.

File: flux/server.py [hunk 4]
```diff
@@ -2958,6 +2958,557 @@ async def schedule_history(
...
+                        identity,
+                        f"workflow:{namespace}:{wf_info.name}:read",
+                    ):
+                        raise HTTPException(
+                            status_code=403,
+                            detail=f"Permission denied: requires 'workflow:{namespace}:{wf_info.name}:read'",
+                        )
+
+                manager = ContextManager.create()
+                context = manager.get(execution_id)
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3566]: Fixed in 29859c7. Added null check — returns 404 when execution context not found.

File: flux/server.py [hunk 5]
```diff
@@ -2958,6 +2958,557 @@ async def schedule_history(
...
+
+                manager = ContextManager.create()
+                ctx = manager.get(execution_id)
+
+                if ctx is None:
+                    raise HTTPException(
+                        status_code=404,
+                        detail=f"Execution context with ID {execution_id} not found.",
+                    )
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3437]: Fixed in 29859c7. Added ownership verification — execution must belong to the resolved workflow, otherwise 404.

File: flux/service_resolver.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            for wf in self._catalog.all(namespace=ns):
+                candidates.setdefault(wf.name, []).append(wf)
+
+        for ref in service.workflows:
+            ns, name = resolve_workflow_ref(ref)
+            wf = self._catalog.get(ns, name)
+            if wf:
+                existing = candidates.get(wf.name, [])
+                if not any(e.namespace == wf.namespace for e in existing):
+                    candidates.setdefault(wf.name, []).append(wf)
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_resolver.py:51]: Fixed. Resolver now wraps catalog.get() in try/except WorkflowNotFoundError — missing workflows are skipped instead of breaking resolution.

File: flux/server.py [hunk 7]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
+            try:
+                if mode not in ["sync", "async", "stream"]:
+                    raise HTTPException(
+                        status_code=400,
+                        detail="Invalid mode. Use 'sync', 'async', or 'stream'.",
+                    )
+
+                resolver = ServiceResolver(WorkflowCatalog.create(), ServiceStore())
+                wf_info = resolver.find(service_name, workflow_name)
+                namespace = wf_info.namespace
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3281]: Fixed. Added except CollisionError → 409 Conflict to service_run_workflow.

File: flux/server.py [hunk 8]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3411]: Fixed. Added except CollisionError → 409 Conflict to service_resume_workflow.

File: flux/server.py [hunk 9]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
+            detailed: bool = False,
+            identity: FluxIdentity = Depends(get_identity),
+        ):
+            from flux.service_resolver import ServiceResolver, WorkflowNotInServiceError
+            from flux.service_store import ServiceNotFoundError, ServiceStore
+
+            try:
+                resolver = ServiceResolver(WorkflowCatalog.create(), ServiceStore())
+                wf_info = resolver.find(service_name, workflow_name)
+                namespace = wf_info.namespace
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3553]: Fixed. Added except CollisionError → 409 Conflict to service_workflow_status.

File: flux/server.py [hunk 10]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
+                            detail=f"Permission denied: requires 'workflow:{namespace}:{wf_info.name}:read'",
+                        )
+
+                manager = ContextManager.create()
+                context = manager.get(execution_id)
+                if context is None:
+                    raise HTTPException(
+                        status_code=404,
+                        detail=f"Execution '{execution_id}' not found",
+                    )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3571]: Fixed. Added ownership verification to status endpoint — checks workflow_namespace and workflow_name match the resolved workflow.

File: flux/server.py [hunk 11]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
+                dto = ExecutionContextDTO.from_domain(context)
+                ctx_dict = dto.model_dump() if hasattr(dto, "model_dump") else dto.dict()
+                return _map_service_response(
+                    ctx_dict,
+                    service_name,
+                    workflow_name,
+                    "sync",
+                    detailed,
+                )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py]: Fixed. Status endpoint now returns execution context summary directly instead of _map_service_response, always including state field.

File: flux/server.py [hunk 12]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
+                    detail="Service name is required and must be a string",
+                )
+            try:
+                store = ServiceStore()
+                svc = store.create(
+                    name=name,
+                    namespaces=body.get("namespaces", []),
+                    workflows=body.get("workflows", []),
+                    exclusions=body.get("exclusions", []),
+                )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py]: Fixed. create_service wraps request.json() in try/except JSONDecodeError → 400, validates namespaces/workflows/exclusions are lists.

File: flux/server.py [hunk 13]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
+                store = ServiceStore()
+                svc = store.update(
+                    name=service_name,
+                    add_namespaces=body.get("add_namespaces"),
+                    add_workflows=body.get("add_workflows"),
+                    add_exclusions=body.get("add_exclusions"),
+                    remove_namespaces=body.get("remove_namespaces"),
+                    remove_workflows=body.get("remove_workflows"),
+                    remove_exclusions=body.get("remove_exclusions"),
+                )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py]: Fixed. update_service same JSON parsing + type validation fix.

File: flux/service_proxy.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+        except httpx.ConnectError:
+            return JSONResponse(
+                status_code=502,
+                content={"detail": "Cannot connect to Flux server"},
+            )
+        except httpx.TimeoutException:
+            return JSONResponse(status_code=504, content={"detail": "Flux server timeout"})
+
+        return JSONResponse(status_code=response.status_code, content=response.json())
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py:169]: Fixed. Standalone proxy rejects mode=stream with 400. Only sync and async supported.

File: flux/server.py [hunk 15]
```diff
@@ -2958,6 +2958,577 @@ async def schedule_history(
...
+
+        @api.post("/services/{service_name}/{workflow_name}")
+        @api.post("/services/{service_name}/{workflow_name}/{mode}")
+        async def service_run_workflow(
+            service_name: str,
+            workflow_name: str,
+            input: Any = Body(None),
+            mode: str = "async",
+            detailed: bool = False,
+            version: int | None = None,
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3262]: Fixed. Default mode changed from async to sync in service_run_workflow and service_resume_workflow. E2E tests updated to use explicit /async path.

File: flux/service_proxy.py [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+            self._endpoints = endpoints
+            self._cache_time = time.monotonic()
+
+    async def resolve(self, workflow_name: str) -> WorkflowRef:
+        if workflow_name not in self._endpoints:
+            self._cache_time = 0.0
+            await self._refresh_cache()
+        if workflow_name not in self._endpoints:
+            raise KeyError(f"Workflow '{workflow_name}' not found in service '{self.service_name}'")
+        return self._endpoints[workflow_name]
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py:65]: Fixed. _refresh_cache wraps raise_for_status() in try/except HTTPStatusError — 404 converts to KeyError, others to ConnectionError.

File: flux/server.py [hunk 17]
```diff
@@ -2958,6 +2958,641 @@ async def schedule_history(
...
+            except (JSONDecodeError, ValueError):
+                raise HTTPException(status_code=400, detail="Invalid JSON body")
+
+            name = body.get("name")
+            if not name or not isinstance(name, str):
+                raise HTTPException(
+                    status_code=400,
+                    detail="Service name is required and must be a string",
+                )
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:2992]: Fixed. Added regex validation — service names must be lowercase alphanumeric + hyphens. Returns 400 if invalid.

File: flux/server.py [hunk 18]
```diff
@@ -2958,6 +2958,641 @@ async def schedule_history(
...
+            for field in ("namespaces", "workflows", "exclusions"):
+                val = body.get(field, [])
+                if not isinstance(val, list):
+                    raise HTTPException(
+                        status_code=400,
+                        detail=f"'{field}' must be a list",
+                    )
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3005]: Fixed. List items are now validated as strings in create_service.

File: flux/server.py [hunk 19]
```diff
@@ -2958,6 +2958,641 @@ async def schedule_history(
...
+            try:
+                store = ServiceStore()
+                svc = store.update(
+                    name=service_name,
+                    add_namespaces=body.get("add_namespaces"),
+                    add_workflows=body.get("add_workflows"),
+                    add_exclusions=body.get("add_exclusions"),
+                    remove_namespaces=body.get("remove_namespaces"),
+                    remove_workflows=body.get("remove_workflows"),
+                    remove_exclusions=body.get("remove_exclusions"),
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3154]: Fixed. List items validated as strings in update_service for all add_*/remove_* fields.

File: flux/server.py [hunk 20]
```diff
@@ -2958,6 +2958,641 @@ async def schedule_history(
...
+        # ===========================================
+
+        def _service_detailed(ctx_dict, service_name, workflow_name, **extra):
+            result = {
+                "execution_id": ctx_dict.get("execution_id"),
+                "state": ctx_dict.get("state"),
+                "output": ctx_dict.get("output"),
+                "namespace": ctx_dict.get("workflow_namespace"),
+                "workflow": ctx_dict.get("workflow_name"),
+            }
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3209]: Intentional — we use `output` not `data` for consistency. PR description updated to match.

File: flux/service_proxy.py [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+        self.cache_ttl = cache_ttl
+        self._client = httpx.AsyncClient(base_url=self.server_url, timeout=30.0)
+        self._lock = asyncio.Lock()
+        self._endpoints: dict[str, WorkflowRef] = {}
+        self._cache_time: float = 0.0
+
+    async def _refresh_cache(self) -> None:
+        async with self._lock:
+            if self._endpoints and (time.monotonic() - self._cache_time) < self.cache_ttl:
+                return
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py:34]: Fixed. Cache timestamp uses None sentinel instead of 0. Empty endpoint map no longer triggers repeated refreshes.

File: flux/service_proxy.py [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+            return JSONResponse(
+                status_code=400,
+                content={"detail": "Standalone proxy only supports 'sync' and 'async' modes"},
+            )
+
+        try:
+            ref = await proxy.resolve(workflow_name)
+        except KeyError as e:
+            return JSONResponse(status_code=404, content={"detail": str(e)})
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py:141]: Fixed. run_workflow catches ConnectionError from proxy.resolve() and returns 502.

File: flux/service_proxy.py [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+            auth_headers["authorization"] = request.headers["authorization"]
+
+        params: dict[str, str] = {}
+        if detailed:
+            params["detailed"] = "true"
+        if version:
+            params["version"] = version
+
+        body = await request.json() if await request.body() else None
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py]: Fixed. Body parsing wrapped in try/except, returns 400 on invalid JSON.

File: flux/service_proxy.py [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+        request: Request,
+        workflow_name: str,
+        execution_id: str,
+        mode: str = Query("sync"),
+    ):
+        try:
+            ref = await proxy.resolve(workflow_name)
+        except KeyError as e:
+            return JSONResponse(status_code=404, content={"detail": str(e)})
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py:186]: Fixed. resume_workflow catches ConnectionError from proxy.resolve() and returns 502.

File: flux/service_proxy.py [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+        path = f"/workflows/{ref.namespace}/{ref.name}/resume/{execution_id}/{mode}"
+        auth_headers = {}
+        if "authorization" in request.headers:
+            auth_headers["authorization"] = request.headers["authorization"]
+
+        body = await request.json() if await request.body() else None
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py]: Fixed. Resume body parsing wrapped in try/except, returns 400 on invalid JSON.

File: flux/server.py [hunk 26]
```diff
@@ -2958,6 +2958,641 @@ async def schedule_history(
...
+                    detail=f"Error resuming workflow via service: {str(e)}",
+                )
+
+        @api.get("/services/{service_name}/{workflow_name}/status/{execution_id}")
+        async def service_workflow_status(
+            service_name: str,
+            workflow_name: str,
+            execution_id: str,
+            detailed: bool = False,
+            identity: FluxIdentity = Depends(get_identity),
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3541]: Fixed. Removed unused `detailed` parameter from service_workflow_status.
